### ISSUE TITLE
Playback wont work if i skip the first frames

### ISSUE THREAD

**Issue by thorleifjaocbsen** [mbebenita/Broadway#175]
Hello.

I have no idea on how this works, but I stream live data from Raspberry PI H264 Camera using the "raspivid" command. I stream the data to a socket where Broadway reads each package spereated by NALseperator (0, 0, 0, 1). 

But if i jump into the "stream" right in the middle. Say i skip the 10 first packages. Then it wont display anything. Is there a way to "bypass" this? So I dont have to restart the streaming and send the first 10 packages every time? 

Is this a bug or how H264 works?

**Comment by soliton4** [comment]
the first frame is usually a combinagion of the stream configuration (dimensions + type of stream info) and what is called an i-frame. 
which is a complete picture.
the following frames are p-frames
see:
https://en.wikipedia.org/wiki/Inter_frame

h264 has some ability to "recover" from lost stream data, but the initial stream configuration  is neccessary.

if your stream i allways made up of the same type and dimensions, you could make a hack where you safe the first frame of a stream in your program and always feed that to broadway first. feeding the next frames should at least display something and after a while the stream might recover

hope that helped
have fun

**Comment by thorleifjaocbsen** [comment]
Seems like that is correct, for Raspberry PI Camera I can see the first two frames and send them manually to restore it.

```
0x27 = 01 00111 = type 7    Sequence parameter set (B-frame)
0x28 = 01 01000 = type 8    Picture parameter set (B-frame)
```

But how would you go to send H264 data from /dev/video2 (usb hardware h264 camera) I've tried everything I could possible think of to get the data and output it. The latest I found was running ffmpeg

`ffmpeg -i /dev/video2 -c copy -f rawvideo pipe:1`

This sends data but it never sends the 0x27 and 0x28, any idea on what might be missing?

**Comment by soliton4** [comment]
you dont want b-frames as they are not part of restricted base profile which is the only profile supported by broadway at this moment.
the frame type byte is not allways the same as it not only contains the frame type information but also some additional flags. if you want to read up on this i encurage you to study h264 bitstream.

i also dont know how to access your video on your unix system. but if you get ffmpeg to output it you can basically use the example in this project -> homepage to encode it in a complatilbe way.

just safe the complete first frame and you should be fine. i mean ... relatively fine as your stream will still look quite distorted for a while.

**Comment by thorleifjaocbsen** [comment]
![image](https://user-images.githubusercontent.com/8550684/41372273-115fc834-6f4d-11e8-8c3f-db820c20df48.png)

Gonna read a bit about h264, this is the output I'm sending to the parser, it never shows any sign of image so something must be missing / wrong. (Thats from the very start)

**Comment by thorleifjaocbsen** [comment]
After a long day searching I finally got video by using the paramters from the demo on this page.

Using them i get a speed of 0.2x on ffmpeg.

But when changing  vcodec from libx264 to "copy" I get a speed of 1x which is realtime.

The problem is that this wont play in Broadway!

I've made a H264 file from that ffmpeg command which does not play for me. here: https://github.com/thorleifjaocbsen/NodeROV_Client/blob/develop/test.h264?raw=true

```ffmpeg -i /dev/video1 -r 30000/1001 -vcodec copy -pass 1 -coder 0 -bf 0 -flags -loop -wpredp 0 -f h264 pipe:1 -y```

What could be wrong with the h264 data from the camera since it discards it?
